### PR TITLE
Compact serialization improvements for enums and serializable types

### LINKED ISSUE #19314: Enum support for Compact Serializer
Currently when object serialized has enum fields (or it's an enum itself), an error is thrown:

> NoSuchMethodException: com.myexample.MyEnum.<init>()

A custom serializer must be made for each possible enum, which is not very user-friendly. Nice enums are core Java features, it would be great to make them serializable by default.

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Fixes #19314, #19323

Adds serialization of enum types and serializable types (without the need to manually register each type).

Breaking changes (list specific methods/types/messages):
* serialized form: format of Enums in Compact serializer
* right now if type is Serializable, then default Java serialization has higher priority - should be Compact Serialization if possible
* no need for no-arg constructor anymore, we can use objenesis like in other parts of Hazelcast and libs like Kryo

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [ ] Request reviewers if possible
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 1]
```diff
@@ -152,7 +169,8 @@ public Object read(@Nonnull CompactReader reader) throws IOException {
     @Nonnull
     private Object createObject(Class associatedClass) {
         try {
-            return ClassLoaderUtil.newInstance(associatedClass.getClassLoader(), associatedClass);
+            ObjectInstantiator instantiator = objectInstantiators.computeIfAbsent(associatedClass, objenesis::getInstantiatorOf);
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:172]: We don't allow dependencies at hazelcast. I would use objenesis as well if that is not the case.

**Comment by sancar** [comment]
Hi @TomaszGaweda, 
I would like to separate Enum support to two topics :
1- When enum is the root object. Example, user is trying to add an Enum to IMap directly. map.put(key, anEnum);
This could be a good addition but it should not be the concern of the Compact format. This should be added to the SerializationService itself ( `SerializationService.serializerFor` should return an Enum serializer ). 
If you feel up-to-it, please send a separate pr.  This could be a good addition.

2- When enum is the field of a Compact Serialized object. 
This requires much broader support than what is implemented here.
Actually, we have discussed this and decided not to add for now. Meaning, if need arises from community or our customers, we can add it in the future. 
The reason that we postponed it is as follows:
When we add a type to the Compact Serialization necessary methods needs to be added to CompactReader,CompactWriter, GenericRecord and GenericRecordBuilder. 
Imagine the method on those as follows:

````
void CompactWriter.writeEnum( String fieldName,  Object enum );
Object CompactReader.readEnum(String fieldName);
```

So far so good, then we want to add it to GenericRecord

```
? GenericRecord.readEnum(String fieldName)
```

Here we have the problem of what to do with the return type. It can not be `Object`. GenericRecord interface is used when user classes are not around. That means that we need to assume that users Enum class might not be available on the classpath as well. So we can not load the class and construct the Enum object. (Note that Similar problem exists on GenericRecordBuilder as well)
It can return `String` or `int`(for the ordinal). Then it does not really different than readInt or readString. User can do the same thing via existing classes. There is nothing readEnum method brings. But we may want to add it for the sake of completeness anyway. 

After you send the pr, we started to discuss if we want to add this support only to ReflectiveSerializer and leave the rest. In this case, a user has to know that it will use GenericRecord.readString for the enum fields if need be. This also is not straight-forward but we could not came to a decision yet.

Your ideas are welcome, but be patient about this pr, as we need to discuss and decide as a team internally. 
Thanks for the contribution and interest.

**Comment by TomaszGaweda** [comment]
Thanks for the feedback :) Sure, I know it may take more time, better discuss now than do breaking changes later.
About the demand: yes, it's quite typical that entities have enums in them, eg. various statuses. Enum values may be sent via distributed invocation to change e.g. some internal system state (case for us). All changes in PR were made to make our project work with Compact Serializers (now works fine with SubZero and Kryo on 4.x). Nothing urgent, we may stick with Kryo when 5.x is released, but I think real-world feedback may help in internal discussions :) In case of any questions, please let me know. I'll wait till decisions are made.

**Comment by mdumandag** [comment]
Hi @TomaszGaweda . 
After an internal discussion, we wanted to support enum fields for the reflective compact serializer (zero-config use case).
So, if you have some time, we could try to update this PR to do that.

I have reviewed the PR, and I want to offer another way of doing things, so that, it would be more aligned with other field types and we won't need an extra dependency.

We want to serialize enum types as Strings, as you did, but instead of treating it as a special kind of class and using read/writeObject to serialize it, we can simply add another `else if` branch to the `createFastReadWriteCaches` method. In this branch, we could check if the type of field we have is an enum, by using `Class#isEnum` method. Then, if it is an enum, we could get the value and construct the reader/writer cache elements similar to others. In the reader cache registration, we could read the string value of the enum member from the reader. Then, if it is null, we could set it as null, if it is not, we can make use of `Enum#valueOf` to get the correct enum member from the string value. The writer cache registration should be similar, get the value, if it is null, write null, if not, write the name of the enum member.

We also need to add the same check to the block starting with `else if (type.isArray())` so that we could also support arrays of enums.

Then, if you could add a test that verifies we can reflectively serialize classes containing enum and enum array values, that would be great.

What do you think about this approach?

**Comment by TomaszGaweda** [comment]
Thanks @mdumandag!

Looks good, just one note: 
> it would be more aligned with other field types and we won't need an extra dependency.

The dependency was to be able to serialize classes without public, empty constructor. For most data classes the requirement soundsd reasonable (after all, JPA etc also requires it), but it becomes problematic when you want to eg. send some distributed computation and the lambda has constructor with some parameters. Sure, it can be reworked to setters, but it would look very ugly :P Using constructor with parameters allows for cleaner code in current Hazelcast. 

What's the reasoning with "no dependencies"? Isn't it already depending on FasterXML Jackson, Classgraph etc?

**Comment by mdumandag** [comment]
@TomaszGaweda I asked the question to my colleagues and the answer was we try to avoid dependencies as long as we can :smile: 

We try to keep the size of the JARs as small as possible and we would like to avoid any possible security vulnerabilities that the extra dependency might introduce in the future.

I agree that being able to serialize classes without an empty constructor is a good enhancement. Maybe, we can require an empty public constructor for now, create an issue to track that enhancement, and if there are enough requests we can try to implement it, possibly using the dependency you mentioned.

**Comment by TomaszGaweda** [comment]
That sounds like a good idea :) 
I'll make changes in this PR today/tomorrow

**Comment by TomaszGaweda** [comment]
Changes are ready, but I messed up something during rebase and how it shows way too many changed files. I'll fix the branch tomorrow

**Comment by TomaszGaweda** [comment]
Closed in favor of #19488, were I applied only necessary changes

**Pull request closed without merging**